### PR TITLE
Add `Map` service combinator

### PR DESCRIPTION
**Pull request by LucioFranco**
Signed-off-by: Lucio Franco <luciofranco14@gmail.com>

### CODE REVIEW

**Comment by LucioFranco** [comment]
@hawkw I like that, will make the change and add the other one. 👍

**Comment by jonhoo** [comment]
This also closes https://github.com/tower-rs/tower/issues/272 I believe?

File: tower/tests/builder.rs [hunk 1]
```diff
@@ -22,6 +22,7 @@ async fn builder_service() {
         .layer(RateLimitLayer::new(5, Duration::from_secs(1)))
         .layer(RetryLayer::new(policy))
         .layer(BufferLayer::new(5))
+        .map(|r: &'static str| r.to_string())
```
Comment (jonhoo) [inline tower/tests/builder.rs:25]: I don't think this actually tests that this call changes the type. Would be good to have that in there somewhere?

**Comment by jonhoo** [comment]
Approving subject to @hawkw's request and the one comment I made about the test :)

File: tower/src/util/map.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    fn call(&mut self, req: R1) -> Self::Future {
+        let req = (self.f)(req);
+        self.inner.call(req)
+    }
+}
+
+/// A map layer that takes some clone `FnMut` and will apply it
+/// to each inner service.
```
Comment (davidbarsky) [inline tower/src/util/map.rs:39]:
```suggestion
/// A [`tower::Layer`] that accepts  a [`Clone`]able [`FnMut`] and applies it
/// to each inner service.
///
/// [`tower::Layer`]: [path-tbd. please find and fix!]
/// [`Clone`]: https://doc.rust-lang.org/std/clone/trait.Clone.html
/// [`FnMut`]: https://doc.rust-lang.org/std/ops/trait.FnMut.html
```

File: tower/src/util/map.rs [hunk 3]
```diff
@@ -0,0 +1,61 @@
+use std::task::{Context, Poll};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// A service to map from one request type to another.
```
Comment (davidbarsky) [inline tower/src/util/map.rs:5]:
```suggestion
/// A [`tower::Service`] which maps from one request type to another.
///
/// [`tower::Service`]: [../../src/trait.Service.html]
```

Link might be wrong, though! Render this locally first.

File: tower/src/util/map.rs [hunk 4]
```diff
@@ -35,8 +37,12 @@ where
     }
 }
 
-/// A map layer that takes some clone `FnMut` and will apply it
+/// A [`tower::Layer`] that accepts  a [`Clone`]able [`FnMut`] and applies it
 /// to each inner service.
+///
+/// [`tower::Layer`]: [path-tbd. please find and fix!]
```
Comment (jonhoo) [inline tower/src/util/map.rs:43]: Just leaving a reminder comment for this.

File: tower/src/util/map/response.rs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    fn clone(&self) -> Self {
+        MapResponse {
+            inner: self.inner.clone(),
+            f: self.f.clone(),
+        }
+    }
+}
+
+/// A [`tower::Layer`] that accepts  a [`Clone`]able [`Fn`] and applies it
+/// to each inner service.
```
Comment (davidbarsky) [inline tower/src/util/map/response.rs:64]: Is this now an `Fn` rather than an `FnMut`? If so, fix the docs?

File: tower/src/util/map/request.rs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    fn call(&mut self, req: R1) -> Self::Future {
+        let req = (self.f)(req);
+        self.inner.call(req)
+    }
+}
+
+/// A [`tower::Layer`] that accepts  a [`Clone`]able [`Fn`] and applies it
+/// to each inner service.
```
Comment (davidbarsky) [inline tower/src/util/map/request.rs:41]: Is this now an `Fn` rather than an `FnMut`? If so, fix the docs?

**Comment by LucioFranco** [comment]
@davidbarsky they are both `Fn` so I think the docs are fine?

File: tower/src/util/map/request.rs [hunk 7]
```diff
@@ -37,11 +37,12 @@ where
     }
 }
 
-/// A [`tower::Layer`] that accepts  a [`Clone`]able [`Fn`] and applies it
-/// to each inner service.
+/// A [`tower::Layer`] that accepts  a [`Fn`] and applies it
+/// to each inner service. Note that [`Fn`] will implement
+/// `Clone` if all the values it captures implements either `Copy`
+/// or `Clone`.
```
Comment (davidbarsky) [inline tower/src/util/map/request.rs]:
```suggestion
/// A [`tower::Layer`] that accepts an [`Fn`] and applies it
/// to each inner service. Note that [`Fn`] will implement
/// `Clone` _if_  all the captured values implement `Copy`
/// or `Clone`.
```

File: tower/src/util/map/request.rs [hunk 8]
```diff
@@ -37,11 +37,12 @@ where
...
 ///
 /// [`tower::Layer`]: ../trait.Layer.html
-/// [`Clone`]: https://doc.rust-lang.org/std/clone/trait.Clone.html
```
Comment (davidbarsky) [inline tower/src/util/map/request.rs]: Why remove this?

File: tower/src/util/map/request.rs [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+/// A [`tower::Layer`] that accepts  a [`FnMut`] and applies it
+/// to each inner service. Note that [`FnMut`] will implement
+/// [`Clone`] _if_ all the values it captured implement either `Copy`
+/// or [`Clone`].
+///
+/// [`tower::Layer`]: ../trait.Layer.html
+/// [`FnMut`]: https://doc.rust-lang.org/std/ops/trait.FnMut.html
+/// [`Clone`]: https://doc.rust-lang.org/std/clone/trait.Clone.html
```
Comment (davidbarsky) [inline tower/src/util/map/request.rs:47]:
```suggestion
/// A [`tower::Layer`] that accepts an [`FnMut`] and applies it
/// to each inner service. Note that [`FnMut`] will implement
/// [`Clone`] _if_ all the captured values implement either [`Copy`]
/// or [`Clone`].
///
/// [`tower::Layer`]: ../trait.Layer.html
/// [`FnMut`]: https://doc.rust-lang.org/std/ops/trait.FnMut.html
/// [`Clone`]: https://doc.rust-lang.org/std/clone/trait.Clone.html
/// [`Copy`]: https://doc.rust-lang.org/std/marker/trait.Copy.html
```

File: tower/src/util/map/response.rs [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    future::Future,
+    marker::PhantomData,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// A [`tower::Service`] to map from one response type to another.
```
Comment (davidbarsky) [inline tower/src/util/map/response.rs:12]:
```suggestion
/// A [`tower::Service`] which maps from one response type to another.
```

File: tower/src/util/map/response.rs [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    fn clone(&self) -> Self {
+        MapResponse {
+            inner: self.inner.clone(),
+            f: self.f.clone(),
+        }
+    }
+}
+
+/// A [`tower::Layer`] that accepts  a [`Clone`]able [`Fn`] and applies it
+/// to each inner service.
```
Comment (davidbarsky) [inline tower/src/util/map/response.rs:64]:
```suggestion
/// A [`tower::Layer`] that accepts a [`Clone`]able [`Fn`] and applies it
/// to each inner service.
```

File: tower/src/util/map/request.rs (hunk 7 above)
Comment (hawkw) [reply tower/src/util/map/request.rs]: (or if it's a `fn`, of course)

**Pull request merged**